### PR TITLE
adapter: Instrument adapter completion

### PR DESCRIPTION
**Pull request by felder-cl**
Instrument the completion status of the adapters and label them by type.  The
Job Spec ID is included aswell.

Should execution time be added as well?  This would help monitor for long
running or stuck adapters.

Also:

 - adapters: Make TaskType inspectable

   Support logic on the TaskType of the adapter by introducing the method
   TaskType() to the Adapter interface.  This will be usefull for logging and
   metrics.

https://www.pivotaltracker.com/n/projects/2129823/stories/171640595

### CODE REVIEW

**Comment by felder-cl** [comment]
@jleeh @j16r what do you think about the labeling? I've been out of sync with work on prometheus in the codebase, so not sure what's kosher.

https://github.com/smartcontractkit/chainlink/pull/2477/files#diff-7b92f2153b0cd04be9d5849fa55618f4R24

File: core/services/run_executor.go [hunk 1]
```diff
@@ -121,5 +132,7 @@ func (re *runExecutor) executeTask(run *models.JobRun, taskRun *models.TaskRun)
 
 	input := *models.NewRunInput(run.ID, data, taskRun.Status)
 	result := adapter.Perform(input, re.store)
+	promAdapterCallsVec.WithLabelValues(run.ID.String(), string(adapter.TaskType()), string(result.Status())).Inc()
```
Comment (felder-cl) [inline core/services/run_executor.go]: this should be Job Spec ID instead of Job Run ID

**Comment by DavidCurrie1113** [comment]
Has nothing to do with you Mexican contract holder trying to get people to
fence fake BNB token and other tokens on binance Bank so I don't want
nothing anything to do with them for him as far as the token contract goes
it's b******* anyway he manages mutual's for people in Nova Scotia and
their elderly if it's you you can go f*** yourself

File: core/services/run_executor.go (hunk 1 above)
Comment (felder-cl) [reply core/services/run_executor.go]: done

File: core/services/run_executor.go [hunk 2]
```diff
@@ -121,5 +132,7 @@ func (re *runExecutor) executeTask(run *models.JobRun, taskRun *models.TaskRun)
 
 	input := *models.NewRunInput(run.ID, data, taskRun.Status)
 	result := adapter.Perform(input, re.store)
+	promAdapterCallsVec.WithLabelValues(run.JobSpecID.String(), string(adapter.TaskType()), string(result.Status())).Inc()
```
Comment (se3000) [inline core/services/run_executor.go]: Won't this count get jacked up for the `PendingConfirmations`, which will happen multiple times per run? Maybe we should decrement the previous state?

Comment (felder-cl) [reply core/services/run_executor.go]: I confirmed w/ @j16r .  These metrics are counted over a timeline so we are tracking changes in the system as opposed to the state of the system.

Comment (j16r) [reply core/services/run_executor.go]: As is this design would count state transitions, so pending confirmations may end up counting more than the number of runs.

If we only wanted to count the total number of jobs that had experienced a PendingConfirmations state, you could subtract whenever a PendingConfirmations -> X state transition had been made.

TBH I'd have both!

**Comment by felder-cl** [comment]
> Add pendingconfirmations metrics for total job count has matched state vs current design that counts total transitions.

Could you please explain? To confirm, do you mean a gauge or a counter to track the number of pending as well as state transitions?

**Comment by j16r** [comment]
Yeah, sorry if it's not clear.

I think there is a case for two sets of counters!

  1) For all state transitions - just increment a counter whenever a new state is applied to a run
  2) For the total states - if a run observes the X state multiple times, only report it once

Not sure how to implement the latter though, as I suspect Steve's suggestion won't work. Will have to think about it more.

**Comment by felder-cl** [comment]
I'm personally unsure whether prometheus should have state exported to it like this.  My assumptions were always just to monitor state transitions.  I believe this may be out of scope though for the current PR.  Ok to proceed?

**Pull request merged**